sm/plugin_sm/src/plugin.rs

```diff
@@ -1,43 +1,64 @@
+use crate::logged_command::LoggedCommand;
 use async_trait::async_trait;
 use json_sm::*;
-use std::{
-    iter::Iterator,
-    path::PathBuf,
-    process::{Output, Stdio},
-};
-use tokio::process::Command;
+use std::{iter::Iterator, path::PathBuf, process::Output};
+use tokio::fs::File;
+use tokio::io::BufWriter;
 
 #[async_trait]
 pub trait Plugin {
-    async fn prepare(&self) -> Result<(), SoftwareError>;
-    async fn install(&self, module: &SoftwareModule) -> Result<(), SoftwareError>;
-    async fn remove(&self, module: &SoftwareModule) -> Result<(), SoftwareError>;
-    async fn finalize(&self) -> Result<(), SoftwareError>;
-    async fn list(&self) -> Result<Vec<SoftwareModule>, SoftwareError>;
-    async fn version(&self, module: &SoftwareModule) -> Result<Option<String>, SoftwareError>;
-
-    async fn apply(&self, update: &SoftwareModuleUpdate) -> Result<(), SoftwareError> {
+    async fn prepare(&self, logger: &mut BufWriter<File>) -> Result<(), SoftwareError>;
```

Comment:
Instead of taking the logger explicitly on each command, could it be passed in just once in the constructor? That would have made this log introduction less invasive across all these files, I feel.

Reply:
Agree, it's used in all the functions, can't we make this as part of the plugin trait?

Reply:
This cannot be done directly.

* A `Plugin` instance is used to execute *all* the commands on some underlying plugin, while we need a fresh log file for each software operation request received from the cloud.
* The log file is not restricted to a single plugin. We need to log all the commands from all the plugins that have been invoked for a software operation request.

That could be done with a struct that wraps a plugin with a logger and that implements the `Plugin` trait. This would not be enough because one needs then a way to capture the output of the inner plugin.

I think this extra complexity is not worth the cost.
* Yes, there is an extra argument for all the methods.
* But, these methods are used only in the module itself.